components/server/src/util/log-context.ts

```diff
@@ -13,24 +13,33 @@ import { v4 } from "uuid";
 type EnhancedLogContext = LogContext & {
     contextId?: string;
     contextTimeMs: number;
+    contextKind: string;
 };
+
 const asyncLocalStorage = new AsyncLocalStorage<EnhancedLogContext>();
 const augmenter: LogContext.Augmenter = (ctx) => {
     const globalContext = asyncLocalStorage.getStore();
-    const contextTime = globalContext?.contextTimeMs ? Date.now() - globalContext.contextTimeMs : undefined;
-    return {
+    const contextTimeMs = globalContext?.contextTimeMs ? Date.now() - globalContext.contextTimeMs : undefined;
+    const result = {
         ...globalContext,
-        contextTime,
+        contextTimeMs,
         ...ctx,
     };
+    // if its an empty object return undefined
+    return Object.keys(result).length === 0 ? undefined : result;
 };
 LogContext.setAugmenter(augmenter);
 
-export async function runWithContext<T>(context: LogContext, fun: () => T): Promise<T> {
+export async function runWithContext<T>(
+    contextKind: string,
+    context: LogContext & { contextId?: string },
+    fun: () => T,
+): Promise<T> {
     return asyncLocalStorage.run(
         {
             ...context,
-            contextId: v4(),
+            contextKind,
+            contextId: context.contextId || v4(),
```

Comment:
Since the contextId can also be passed from outside for instance when reusing the tracing context id, I would like to keep it in tact so they match. Let's see how this works and if we have situations where we deal with contextIds without knowing it is one, we might come back to this.

Reply:
Fine with coming back later.
Just to clarify: I meant prefixing the id created here like `contextId: context.contextId || `ctxid_${v4()}`. :+1: